### ISSUE TITLE
train datasets

### ISSUE THREAD

**Issue by 104343** [aimagelab/ReT-2#3]
Thank you for your excellent work. May I ask if it would be possible to make the JSONL files used to organize the training dataset publicly available?

**Comment by dcaffo98** [comment]
Hi @104343 

The training data available in [Hugging Face](https://huggingface.co/datasets/aimagelab/ReT-M2KR) comprises many `jsonl` files, one for each dataset in M2KR or M-BEIR, and they must be concatenated to build up the full training dataset.
For convenience, we have uploaded the concatenated version for [M2KR](https://huggingface.co/datasets/aimagelab/ReT-M2KR/blob/main/jsonl/m2kr/train/m2kr_up_train.jsonl) and [M-BEIR.](https://huggingface.co/datasets/aimagelab/ReT-M2KR/blob/main/jsonl/mbeir/train/mbeir_train.jsonl).
Concerning M2KR, the concatenated data is already modified according to ReT [1], that is, we downsampled data for InfoSeek, we upsampled data 2x for E-VQA, and we upsampled data 9x for OK-VQA.

[1] Caffagni, Davide, _et al._ "Recurrence-Enhanced Vision-and-Language Transformers for Robust Multimodal Document Retrieval." In CVPR 2025.

**Comment by 104343** [comment]
Thank you. We would like to know whether the variant of ReT-2 mentioned in the paper, in which the visual and textual backbones are unfrozen during training, corresponds to fully fine-tuning all parameters of the CLIP visual and text encoders, or only partially unfreezing certain layers. We ask this because fully unfreezing all CLIP layers would substantially increase the memory consumption during training.

**Comment by dcaffo98** [comment]
Yes, by _backbone unfrozen_ we mean full fine-tuning of all their parameters. And yes, we know that it gets annoying because of the high memory usage. However, as we built ReT-2 on top of Transformer's `PreTrainedModel`, [it supports gradient checkpointing](https://github.com/aimagelab/ReT-2/blob/3569a585e9e19150c2cb35e2f39092881ff64f34/src/models/ret2/modeling_ret2.py#L160) on the backbones, and that greatly diminishes the memory usage (at the cost of slower training). You can activate checkpointing by passing the `--gradient_checkpointing` flag as a command-line argument to your [training script](https://github.com/aimagelab/ReT-2/blob/main/src/train/train.py).
We always use checkpointing when fine-tuning the backbones. With that enabled, you should be able to fine-tune the CLIP ViT-L/14 backbones with a global batch size of 768-1024 on 4 GPUs of ~64GB each.

**Comment by 104343** [comment]
Thank you very much for your excellent work. May I ask whether you could share the exact training command and parameter settings used for M-BEIR? We are currently encountering some issues with full-parameter fine-tuning and would greatly appreciate your guidance.

**Comment by dcaffo98** [comment]
Can you gently confirm that you can run this script? You can safely remove anything related to SLURM if you are not using it.

NOTE: I have updated both the codebase and the [M-BEIR annotation file](https://huggingface.co/datasets/aimagelab/ReT-M2KR/blob/main/jsonl/mbeir/train/mbeir_train.jsonl). 
```bash
#!/bin/bash
#SBATCH --job-name=train_mbeir
#SBATCH --output=./logs/%x-%j
#SBATCH --error=./logs/%x-%j
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --gpus-per-node=4
#SBATCH --mem=480G
#SBATCH --cpus-per-task=32
#SBATCH --partition=
#SBATCH --qos=
#SBATCH --account=
#SBATCH --time=24:00:00

source activate ret2

# TODO: change this
cd ~/git/ReT-2
export PYTHONPATH=.

# TODO: change these
MBEIR_IMG_ROOT="M-BEIR/mbeir_images"
MBEIR_INST_PATH="M-BEIR/instructions/query_instructions.tsv" # https://huggingface.co/datasets/TIGER-Lab/M-BEIR/blob/main/instructions/query_instructions.tsv


IFS=',' read -r -a nodelist <<<$SLURM_NODELIST
export MASTER_ADDR=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)
export MASTER_PORT=`comm -23 <(seq 5000 6000 | sort) <(ss -Htan | awk '{print $4}' | cut -d':' -f2 | sort -u) | shuf | head -n 1`

srun torchrun --nnodes=$SLURM_NNODES --nproc-per-node=$SLURM_GPUS_PER_NODE --rdzv-endpoint=$MASTER_ADDR --rdzv-id=$SLURM_JOB_NAME --rdzv-backend=c10d \
src/train/train.py \
--seed 42 \
--output_dir "./checkpoints/${SLURM_JOB_NAME}" \
--dataset_train "ReT-M2KR/jsonl/mbeir/train/mbeir_train.jsonl" \
--dataset_image_root "$MBEIR_IMG_ROOT" \
--dataset_sample_mbeir_instructions True \
--dataset_mbeir_instructions_path "$MBEIR_INST_PATH" \
--mbeir_datatask_balance_sampler True \
--remove_unused_columns False \
--dataloader_drop_last True \
--dataloader_num_workers 8 \
--dataloader_persistent_workers False \
--num_train_epochs 20 \
--learning_rate 1e-4 \
--warmup_ratio 0.01 \
--lr_scheduler_type cosine \
--backbone_learning_rate_multiplier 0.05 \
--weight_decay 0.0 \
--per_device_train_batch_size 192 \
--gradient_accumulation_steps 1 \
--save_strategy epoch \
--bf16 \
--logging_steps 5 \
--log_on_each_node False \
--report_to wandb \
--run_name "$SLURM_JOB_NAME" \
--gradient_checkpointing True \
--text_name_or_path "openai/clip-vit-large-patch14" \
--text_frozen False \
--text_layer_strategy 'vit_b' \
--text_global_feats_proj_size 0 \
--vision_name_or_path "openai/clip-vit-large-patch14" \
--vision_frozen False \
--vision_layer_strategy vit_l \
--vision_global_feats_proj_size 0 \
--w_final_size 768 \
--logit_scale_init_value 4.605170249938965 \
--logit_scale_clamp 100.0 \
--hidden_size 1024 \
--dropout_p 0.05 \
--attention_dropout 0.05 \
--input_gate_bias 0.0 \
--forget_gate_bias 0.0 \
```

**Comment by 104343** [comment]
Thank you for providing the script. I will try it.

**Comment by 104343** [comment]
Thank you! I also have a question regarding the M2KR dataset. In the training command, --num_train_epochs is set to 30, while the paper reports 75k steps. After doing some calculations, these two settings do not seem to match. Could you please clarify how the training schedule is actually configured?

**Comment by dcaffo98** [comment]
Hi @104343,

We set `--num_train_epochs=30` to be consistent with the learning rate cosine scheduler employed in ReT-1. Similarly, we stopped training at 75k steps as done in ReT-1 (in ReT-1, that is roughly equivalent to 24h of training).
Training further may benefit some benchmarks, although we found a negative impact on InfoSeek and Ok-VQA.

**Comment by kpark53** [comment]
Hi @dcaffo98, Thank you for the excellent work!

I was wondering if you could share the exact training command and parameter settings used for the M2KR dataset. Additionally, for the base-sized text and vision feature extractors, could you please clarify which CLIP ViT-Base version was used (e.g., CLIP ViT-B/16, openai/clip-vit-base-patch16)?

Thank you in advance for your time and help.

**Comment by dcaffo98** [comment]
Dear @kpark53 ,

Thank you for your interest. We use [CLIP ViT-B/32](https://huggingface.co/openai/clip-vit-base-patch32) as the base vision encoder. 
Concerning M2KR training, you may look at this issue #5 .

Best,
Davide

**Comment by 104343** [comment]
May I ask whether you would be willing to release the training hyperparameters used for M2KR when the backbone was frozen during training? Are the training hyperparameters exactly the same as those used in the unfrozen setting?

**Comment by dcaffo98** [comment]
Dead @104343 
To train with frozen backbones, you can take the script provided in #5, with the following modifications:
1. set `--vision_frozen` and `--text_frozen` to `True`
2. remove `--backbone_learning_rate_multiplier`
3. (optional) turn off gradient checkpointing: `--gradient_checkpointing False`

**Comment by 104343** [comment]
thank you, set --vision_frozen and --text_frozen to True?

**Comment by dcaffo98** [comment]
> thank you, set --vision_frozen and --text_frozen to True?

Yeah, sorry for the typo. I've fixed my comment above.
